JDK-8259499: Handling type arguments from outer classes for inner class in javadoc

hns opened the pull request:
This change adds support for generating HTML links to the type arguments of enclosing classes when creating a link to an inner class. Previously, only a link to the inner class was created and the type arguments were even omitted from the link label.

The new feature to create separate links to the enclosing class and its type arguments is only activated if the enclosing class has type arguments. If the enclosing class is non-generic, the old behavior is preserved to create a single link to the inner class. The reason for this is that a dedicated link to the enclosing class itself provides little benefit, since it can be easily reached via the "Enclosing class" link of the inner class. Also, linking the enclosing type in absence of type arguments makes it hard to see that there are two links and easy to click on the wrong link by mistake.

On the other hand, for type arguments a separate link should be useful since it is often not a "nearby" type. It is also easier to detect the different links than for non-generic nested classes. I came to like this "mixed" solution best after trying several other approaches.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259499](https://bugs.openjdk.java.net/browse/JDK-8259499): Handling type arguments from outer classes for inner class in javadoc


### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**)
 * @orkungdk (no known github.com user name / role) ⚠️ Review applies to b4702d9705dd9633bdb7f202d34280077396f7f9


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/195/head:pull/195` \
`$ git checkout pull/195`

Update a local copy of the PR: \
`$ git checkout pull/195` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/195/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 195`

View PR using the GUI difftool: \
`$ git pr show -t 195`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/195.diff">https://git.openjdk.java.net/jdk17/pull/195.diff</a>

</details>

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:
```diff
@@ -1079,7 +1079,7 @@ public Content seeTagToContent(Element element, DocTree see, TagletWriterImpl.Co
             // Must be a class reference since refClass is not null and refMemName is null.
             if (labelContent.isEmpty()) {
-                if (!refClass.getTypeParameters().isEmpty() && seeText.contains("<")) {
+                if (seeText.contains("<")) {
```
Hmmm, string-based checks are generally something of a red-flag, as compared to more abstract methods on appropriate classes. And, this is a string operation on the output of  `CommentHelper.getText`. Is there a way that we have may this a more robust test ... for example introduce `CommentHelper.isGeneric(ReferenceTree)` or something like that?

hns replied:
I agree something like `CommentHelper.isGeneric(ReferenceTree)` would be preferable, but that method would have to iterate over enclosing types to find any generic types there, just like the `LinkInfo.linkEnclosingTypes(DeclaredType)` method I have added. So I wonder where a shared method for this could go if not to `Utils` (I shudder to think).

jonathan-gibbons replied:
As much as I dislike `Utils` it is probably the correct location in this case.  I think of `Utils` as being javadoc-specific enhancements and variations on the Language Model `Elements` and `Types` interfaces.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/links/LinkInfo.java:
```diff
@@ -114,6 +115,31 @@
+     * separate links
+     */
+    public boolean linkEnclosingTypes(DeclaredType type) {
+        // Only render enclosing types as separate links if they have type arguments that
+        // should be linked. The rationale is that an enclosing class itself is easy to reach
+        // from the documentation of an inner class, so rendering it as separate link would
+        // add little benefit but add considerable noise, while type arguments may not
+        // otherwise be reachable from the documentation of the inner class.
```
Do you render the _enclosing types_ as separate links, or just  the _type arguments of enclosing types_?

Even if the enclosing type has type arguments, your argument about the noise from linking the name of the enclosing type is still valid.

liach replied:
Currently the first. But if someone clicks the class label in front of type arguments, thinking it would lead to the class that defines the type parameters (as in many IDEs) and instead goes to the innermost class, it'd be quite confusing too.

hns replied:
Yes, it is either all type components as separate links, or none.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/links/LinkInfo.java:
```diff
@@ -125,7 +151,10 @@ public Content getClassLinkLabel(BaseConfiguration configuration) {
         } else if (isLinkable()) {
             Content tlabel = newContent();
-            tlabel.add(configuration.utils.getSimpleName(typeElement));
+            tlabel.add(type instanceof DeclaredType dt && linkEnclosingTypes(dt)
```
There's a philosophical discussion to be had, separately, with Joe, about using `instanceof` with Language Model types. since the beginning, i`nstanceof` has been discouraged in favor of using `getKind()` but the reasons for that may have faded, and "instanced patterns" may be enough of a reason to change the guidelines.

jonathan-gibbons reviewed (commented):
What is the type arguments are themselves type variables: e.g. `@see Map<P,Q>`

liach commented on test/langtools/jdk/javadoc/doclet/testGenericTypeLink/pkg1/A.java:
```diff
@@ -60,6 +60,11 @@ public void someMethod(List<? extends Number> list, int i) {}
 
+    /**
+     * @see A<String, java.lang.RuntimeException>.Inner
+     * @see A<A<String, java.lang.RuntimeException>.Inner, A.SomeException>
```
Is this a new feature that we can link with type arguments? Interesting, as https://docs.oracle.com/en/java/javase/16/docs/specs/javadoc/doc-comment-spec.html#see has no mention on usage of type arguments in see links.

liach commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlLinkInfo.java:
```diff
@@ -28,6 +28,7 @@
 import javax.lang.model.element.Element;
 import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.DeclaredType;
```
Extraneous import

liach reviewed (commented):
Thank you for addressing an issue I've reported!

hns commented:
> What is the type arguments are themselves type variables: e.g. `@see Map<P,Q>`

We then render them as links to the defining type (in this case `Map`). This is older behavior that has been there before, and it only works for included types.

hns commented:
I have added a commit to this PR to add a Utils#isGenericType(TypeMirror) method that is used from `LinkInfo`, `LinkFactory` and `HtmlDocletWriter`. I think the code is actually a bit nicer now.

Interestingly, I stumbled over a NPE caused by JDK-8269706 when a generic reference is inherited from an overridden method. This bug existed before the last commit (and I think before this PR, although I didn't test this). Since we do not have the time to fix the root problem for 17, I added a workaround (triple null check in `CommentHelper`) as well as a test for the condition in `TestGenericTypeLink.java`.

jonathan-gibbons commented:
TL;DR: OK

I found this a hard one to review, to understand the implications. Even reading the tests, and the output generated from the test was hard. Eventually, I wrote my own simple demo program, with a couple of classes, one generic, one not, both with inner classes. Then, one has methods to referring to each of the inner classes. This setup allowed me to see the different behavior of "side-by-side" links to inner classes with and without a generic outer class.

The observed behavior seems reasonable/sensible.  When the inner class has a non-generic outer class, you get a single link for the pair.  When the inner class has a generic outer class, the single link is split apart, to give links for the outer class and its type args.   That being said, it's not a slam-dunk win, it would be almost equally reasonable to use the same linking strategy for both generic and non-generic outer classes. I'm OK with the current, let's see if we get comments down the road.

The test was mildly confusing for having a series of `{@link}` tags with no non-white text to separate them.

Related but separate: The combination of `{@linkplain}` with no description, causing the default label to be generated in serif font seems questionable, so much so that I almost posted screenshots and comments on the bug.  Maybe this should be a warning in doclint. Generally, displaying a code signature in serif font seems bad.

Related by separate: The "enclosing class" in the class signature is shown in serif font.

jonathan-gibbons reviewed (approved):
See preceding comments.

hns commented:
/integrate
